### PR TITLE
Automate Protobuf Version Templating

### LINKED ISSUE #24897: grpcio-tools==1.34.0 seg faults on macOS Big Sur, depending on import order
### What version of gRPC and what language are you using?

python (any 3.6+ version), gRPC 1.34.0

### What operating system (Linux, Windows,...) and version?

macOS Big Sur 11.0.1

### What runtime / compiler are you using (e.g. python version or version of gcc)

Any python version 3.6+ will exhibit this issue.

```
~: /usr/bin/python3 --version
Python 3.7.3

~: : gcc --version
Configured with: --prefix=/Library/Developer/CommandLineTools/usr --with-gxx-include-dir=/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/4.2.1
Apple clang version 11.0.0 (clang-1100.0.33.8)
Target: x86_64-apple-darwin20.1.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

### What did you do?

Importing a `_pb2.py` file before importing `grpc` leads to a seg fault.

You can reproduce by using this small shell script in a temp folder:
```
#!/bin/bash -xe

PY=${1:-/usr/bin/python3}

if [[ ! -f $PY ]]; then
    PY=$PY/bin/python
fi

rm -rf venv __pycache__
wget -q https://raw.githubusercontent.com/grpc/grpc/master/test/http2_test/messages_pb2.py -Omessages_pb2.py
virtualenv --python $PY venv
venv/bin/pip install -U pip setuptools wheel
venv/bin/pip install grpcio-tools==1.34.0
venv/bin/python --version
venv/bin/python -c 'import grpc; import messages_pb2'
venv/bin/python -c 'import messages_pb2; import grpc'
```

The 1st import sequence works fine: `import grpc; import messages_pb2`
The reverse import sequence seg faults: `import messages_pb2; import grpc`

### What did you expect to see?

Should not seg fault, regardless of import order.

### What did you see instead?

`Segmentation fault: 11`

### Anything else we should know about your project / environment?

This bug does not occur on Linux, and does not occur with previous version `grpcio-tools==1.33.2` on Big Sur.

### PR DESCRIPTION
**Pull request by gnossen**
~~Fixes #24897~~

~~Since both the `protobuf` and `grpcio-tools` packages include `libprotobuf` as a shared library, an application loading both of them has symbol conflicts that need to be resolved by the operating system's runtime linker. In most cases, the runtime linker uses the first loaded symbol, ignoring the second. In corner cases, such as in MacOS in #24897, some startup functions were run by the version from `grpcio-tools` while the version packaged with `protobuf` was called at runtime. This resulted in corruption from the perspective of the version of `libprotobuf` packaged by the  `protobuf` wheel because internal data structures had changed between the two versions.~~

~~By requiring an exact version, this is no longer an issue. This also brings Python into line with other wrapped languages, including Ruby, C#, and Objective-C.~~ Since updating the Protobuf version has been a tedious manual process in the past, I went ahead and pulled all languages into the templating system so in the future, only one instance of the protobuf version string will have to be changed -- in `build_handwritten.yaml`.

CC @jtattermusch since some of these changes will have to be reconciled with https://github.com/grpc/grpc/pull/24855. Once this is merged, I'll update the "Updating third_party/protobuf submodule in grpc" document accordingly.

CC @yulin-liang since this also touches Obj C files.

**Update:** @lidizheng made the good point that the protobuf Python package has much narrower platform support than we do, so pinning the version would be too restrictive. I've removed that portion of the PR so that this is now just an automation change for updaing the protobuf version. I'll address #24897 in a series of follow-up PRs.

### CODE REVIEW

**Review by jtattermusch — APPROVED** [review]
LGTM for the general idea, but some tests seem to be failing.

**Comment by jtattermusch** [comment]
Btw the `protobuf` and `grpcio-tools` conflict only happens in python because of the weird way the protoc binaries are distributed in python (as a python native extension as opposed to just binaries). FWIW it might be useful to simplify the grpcio-tools package by not duplicating the entire protobuf build (for which we extract the build metadata from protobuf's bazel - which has always been a hack).

File: tools/distrib/python/grpcio_tools/setup.py
Comment (lidizheng) [inline tools/distrib/python/grpcio_tools/setup.py:233]: ProtoBuf is falling behind in recent binary wheel releasing. They don't release for Linux 32bit, 3.8 on Windows and [3.9](https://github.com/protocolbuffers/protobuf/issues/7978). This PR's change might be blocked by them. I would recommend to find another more tolerant solution.

File: templates/grpc.gemspec.template
Comment (lidizheng) [inline templates/grpc.gemspec.template:37]: Just curious, why Ruby needs a shorter version number?

Comment (gnossen) [reply templates/grpc.gemspec.template:37]: Good question. I'm not sure, but the protobuf update doc explicitly called this out. Maybe @apolcyn knows?

Comment (apolcyn) [reply templates/grpc.gemspec.template:37]: @gnossen which doc are you referring to?

If we list a protobuf patch release version here, then this gemspec will only permit the picked up google-protobuf dependency's patch version to change, i.e. the minor version will be pinned. But if we only list the minor version number, then the minor version number will be allowed to move.

Maybe `protobuf_short_version` should be called `protobuf_minor_version` ?

Comment (gnossen) [reply templates/grpc.gemspec.template:37]: I'm referring to go/grpc-third-party-protobuf-update-instructions:
> ruby: update "google-protobuf" version in templates/grpc.gemspec.template  to v3.X (WITHOUT the patch version - 3.x.y is NOT correct) and run tools/buildgen/generate_projects.sh. Check that the gem version is available  at https://rubygems.org/gems/google-protobuf

Technically, the short version includes both major and minor. I would expect a variable called `protobuf_minor_version` to (currently) be "13", not "3.13". How about `protobuf_major_minor_version`?  More verbose, but less confusing IMO.

File: tools/distrib/python/grpcio_tools/setup.py
Comment (gnossen) [reply tools/distrib/python/grpcio_tools/setup.py:233]: Good point. I have some alternative ideas to address this that still require `grpcio-tools` to know what version of `protobuf` it was compiled against. For the moment, I'm just going to remove the `install_requires` change. This PR will still give us some nice automation for the next time we update protobuf, which should be happening roughly every gRPC release.

Comment (lidizheng) [reply tools/distrib/python/grpcio_tools/setup.py:233]: btw, they don't yet release source wheels, so users don't have the option to compile themselves.

**Review by lidizheng — APPROVED** [review]
LGTM if we don't change "install_requires" for now.

File: templates/grpc.gemspec.template
Comment (apolcyn) [reply templates/grpc.gemspec.template:37]: `protobuf_major_minor_version` SGTM

File: tools/buildgen/plugins/expand_version.py
Comment (apolcyn) [inline tools/buildgen/plugins/expand_version.py:131]: nit: change the `-1` index to `2` here to be more explicit about capturing these two numbers?

**Review by apolcyn — APPROVED** [review]

**Comment by gnossen** [comment]
Unrelated failure: https://github.com/grpc/grpc/issues/24937

**Pull request merged**
